src/toolkits/drawing_classifier/drawing_classifier.hpp

```diff
@@ -170,19 +186,80 @@ class EXPORT drawing_classifier: public ml_model_base {
   REGISTER_CLASS_MEMBER_FUNCTION(drawing_classifier::export_to_coreml,
                                  "filename");
 
+  REGISTER_CLASS_MEMBER_FUNCTION(drawing_classifier::init_train, "data",
+                                  "target_column_name", "feature_column_name",
+                                  "validation_data", "options");
+  register_defaults("init_train",
+                    {{"validation_data", to_variant(gl_sframe())},
+                     {"options",
+                      to_variant(std::map<std::string, flexible_type>())}});
+
+  REGISTER_CLASS_MEMBER_FUNCTION(drawing_classifier::perform_training_iteration);
+
   END_CLASS_MEMBER_REGISTRATION
 
  protected:
 
+  // Constructor allowing tests to set the initial state of this class and to
+  // inject dependencies.
+  drawing_classifier(
+      const std::map<std::string, variant_type>& initial_state,
+      std::unique_ptr<neural_net::model_spec> nn_spec,
+      std::unique_ptr<neural_net::compute_context> training_compute_context,
+      std::unique_ptr<data_iterator> training_data_iterator,
+      std::unique_ptr<neural_net::model_backend> training_model)
+      : nn_spec_(std::move(nn_spec)),
+        training_data_iterator_(std::move(training_data_iterator)),
+        training_compute_context_(std::move(training_compute_context)),
+        training_model_(std::move(training_model)) {
+    add_or_update_state(initial_state);
+  }
+
+  // Factory for data_iterator
+  virtual std::unique_ptr<data_iterator> create_iterator(
+      data_iterator::parameters iterator_params) const;
+
+  std::unique_ptr<data_iterator> create_iterator(
+      gl_sframe data, bool is_train,
+      std::vector<std::string> class_labels) const;
+
+  // Factory for compute_context
+  virtual std::unique_ptr<neural_net::compute_context> create_compute_context()
+      const;
+
   // Returns the initial neural network to train
-  virtual std::unique_ptr<neural_net::model_spec> init_model() const;
+  virtual std::unique_ptr<neural_net::model_spec> init_model(
+      bool random_init) const;
+
+  virtual std::tuple<gl_sframe, gl_sframe> init_data(
+      gl_sframe data, variant_type validation_data) const;
+
+  virtual std::tuple<float, float> compute_validation_metrics(
+      size_t num_classes, size_t batch_size);
+
+  virtual void init_table_printer(bool has_validation);
 
+  // Utility code
+  template <typename T>
+  T read_state(const std::string& key) const {
+    return variant_get_value<T>(get_state().at(key));
+  }
 
  private:
   // Primary representation for the trained model.
   std::unique_ptr<neural_net::model_spec> nn_spec_;
-  std::unique_ptr<neural_net::model_backend> training_model_;
+
+  // Primary dependencies for training. These should be nonnull while training
+  // is in progress.
+  gl_sframe training_data_;  // TODO: Avoid storing gl_sframe AND data_iterator.
```

Comment:
why not store the data into your data_iterator?

Reply:
Yeah, since it's already stored in the data iterator parameters, I'll take it out.

Reply:
Resolving per offline discussion.